### PR TITLE
PUBDEV-5852: Fillna Backward Implementation

### PR DESCRIPTION
**Pull request by wendycwong**
This PR completes the work for JIRA: https://0xdata.atlassian.net/browse/PUBDEV-5852?filter=myopenissues

Worked done:
1. Added backward fillna operation to AstFillNA across columns and rows.
2. Added simple fillna operations tests with handcrafted answer to verify AstFillNA operation and also a single thread implementation  in AstFillNATest.java..
3.  Added fillna operations on large frame in AstFillNATest.java.

### CODE REVIEW

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstFillNA.java [hunk 1]
```diff
@@ -54,19 +54,25 @@ public Val apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
...
-    // Not impl yet
-    if (METHOD_BACKWARD.equalsIgnoreCase(method.trim()))
-      throw H2O.unimpl("Backward fillNA method is not yet implemented.");
 
     final int axis = (int) asts[3].exec(env).getNum();
     if (!(Arrays.asList(0,1)).contains(axis))
       throw new IllegalArgumentException("Axis must be 0 for columnar 1 for row");
     final int limit = (int) asts[4].exec(env).getNum();
+
+    if (limit == 0) // fast short cut to do nothing if user set zero limit
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstFillNA.java]: The operation makes no sense if the limit is zero. This is for another Jira, but I'd validate the `limit` argument on the client already or at least on the backend. 

However, in current state, this is valuable here and I'd leave it. I'm just wondering of we should throw an exception and tell the user the operation has no sense at all ? Do we even validate it for limit < 0 ? An assertion for that could be considered here.

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstFillNA.java [hunk 2]
```diff
@@ -98,6 +105,41 @@ public void map(Chunk cs[], NewChunk nc[]) {
       }
     }
   }
+
+  /***
+   * This class was shameless copied from FillForwardTaskRow from Pavel.  We gain speed here by not having
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstFillNA.java]: I'd remove the first sentence. No need for apologies and the FillForwardTaskRow is not written by me at all. 

The **code duplication is ok**, as speed gain is what we want when iterating through chunks.Not really a duplication. In theory, algorithms should be general. In practice, we need them to finish ASAP :)

File: h2o-core/src/test/java/water/rapids/ast/prims/mungers/AstFillNATest.java [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+    } finally {
+      Scope.exit();
+    }
+  }
+
+  /**
+   * Purpose here is to carry out short tests to make sure the code works and the single thread code works.
+   */
+  @Test
+  public void testBackwardMethodRowAll() {
```
Comment (Pscheidl) [inline h2o-core/src/test/java/water/rapids/ast/prims/mungers/AstFillNATest.java]: I like the speed of this solution. Fast tests is what we need.

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 4]
```diff
@@ -3542,7 +3542,7 @@ def build_save_model_DRF(params, x, train, respName):
 
 
 # generate random dataset, copied from Pasha
-def random_dataset(response_type, verbose=True, NTESTROWS=200):
+def random_dataset(response_type, verbose=True, NTESTROWS=200, missing_fraction=0.0, seed=None):
```
Comment (Pscheidl) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: I don't understand the purpose of this change.  Looks like a valuable addition, but is this a leftover ?

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstFillNA.java (hunk 1 above)
Comment (wendycwong) [reply h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstFillNA.java]: I like your suggestion here.  We should throw an exception when limit < 0 and do nothing when limit == 0 at least for now.

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 4 above)
Comment (wendycwong) [reply h2o-py/tests/pyunit_utils/utilsPY.py]: This was written by Pasha originally.  I just copied it over and keep adding to it.  Yes, it was a leftover.

**Comment by michalkurka** [comment]
@Pscheidl please review this one, thank you

**Comment by Pscheidl** [comment]
@michalkurka @wendycwong Approved. Just make sure you really want the ` h2o-py/tests/pyunit_utils/utilsPY.py`file changed.

**Pull request merged**